File: go/webhooks.go
package videofetch

import (
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"net/http"
	"net/url"
	"strconv"
	"strings"
)

// ErrBadSignature is returned when a webhook signature does not verify.
var ErrBadSignature = errors.New("videofetch: signature does not match the payload and secret")

// Additional account-level webhook event names (see WEBHOOK_EVENTS on the server).
const (
	EventQuotaWarning  = "quota.warning"
	EventQuotaExceeded = "quota.exceeded"
	EventBalanceLow    = "balance.low"
	// EventStorageConnectionFailed fires once when a saved connection starts failing
	// with a non-retryable error (auth, permission, missing bucket, …).
	EventStorageConnectionFailed = "storage.connection_failed"
)

// WebhookEvents lists every event an account-level endpoint can subscribe to.
var WebhookEvents = []string{
	EventQueued, EventProcessing, EventCompleted, EventFailed,
	EventQuotaWarning, EventQuotaExceeded, EventBalanceLow, EventStorageConnectionFailed,
}

// WebhookEndpoint is a registered account-level webhook endpoint.
// Secret is the full plaintext value only right after Create; List returns it masked.
type WebhookEndpoint struct {
	ID             string   `json:"id"`
	URL            string   `json:"url"`
	Secret         string   `json:"secret"`
	Events         []string `json:"events"`
	Active         bool     `json:"active"`
	LastDeliveryAt *string  `json:"last_delivery_at"`
	LastStatus     *int     `json:"last_status"`
	FailureCount   int      `json:"failure_count"`
	CreatedAt      *string  `json:"created_at"`
}

// WebhookList is the GET /v1/webhooks response (secrets are masked).
type WebhookList struct {
	Items []WebhookEndpoint `json:"items"`
}

// WebhookTestResult is the POST /v1/webhooks/{id}/test response.
type WebhookTestResult struct {
	Delivered       bool   `json:"delivered"`
	URL             string `json:"url"`
	LastStatus      *int   `json:"last_status"`
	SignatureHeader string `json:"signature_header"`
	SignatureFormat string `json:"signature_format"`
}

// WebhookDeliveryHealth aggregates delivery outcomes for an endpoint.
type WebhookDeliveryHealth struct {
	Total       int     `json:"total"`
	Succeeded   int     `json:"succeeded"`
	Dead        int     `json:"dead"`
	Pending     int     `json:"pending"`
	SuccessRate float64 `json:"success_rate"`
}

// WebhookDelivery is one recorded delivery attempt of an event to an endpoint.
// Deliveries are at-least-once and are not ordered: order by Seq, not CreatedAt.
type WebhookDelivery struct {
	EventID        string  `json:"event_id"`
	Event          string  `json:"event"`
	Seq            int64   `json:"seq"`
	Status         string  `json:"status"` // pending|succeeded|dead
	Attempts       int     `json:"attempts"`
	MaxAttempts    int     `json:"max_attempts"`
	LastStatusCode *int    `json:"last_status_code"`
	LastError      *string `json:"last_error"`
	NextAttemptAt  *string `json:"next_attempt_at"`
	CreatedAt      *string `json:"created_at"`
	DeliveredAt    *string `json:"delivered_at"`
	DownloadID     *string `json:"download_id"`
}

// WebhookDeliveriesResult is the GET /v1/webhooks/{id}/deliveries response.
type WebhookDeliveriesResult struct {
	EndpointID           string                `json:"endpoint_id"`
	Active               bool                  `json:"active"`
	AutoDisabledAt       *string               `json:"auto_disabled_at"`
	ConsecutiveFailures  int                   `json:"consecutive_failures"`
	MaxAttempts          int                   `json:"max_attempts"`
	RetryScheduleSeconds string                `json:"retry_schedule_seconds"`
	Health               WebhookDeliveryHealth `json:"health"`
	Items                []WebhookDelivery     `json:"items"`
}

// ReplayResult is the POST /v1/webhooks/deliveries/{event_id}/replay response.
type ReplayResult struct {
	Queued  bool   `json:"queued"`
	EventID string `json:"event_id"`
}

// WebhooksService manages account-level webhook endpoints. Authentication is by
// API key (no JWT needed), so a backend integration can self-provision callbacks.
// Access via client.Webhooks.
type WebhooksService struct{ client *Client }

// Create registers a new endpoint. With no events it subscribes to every event.
// The returned Secret is the FULL plaintext value — this is the only time the
// server ever returns it, so store it securely.
func (s *WebhooksService) Create(ctx context.Context, hookURL string, events ...string) (*WebhookEndpoint, error) {
	body := map[string]any{"url": hookURL}
	if len(events) > 0 {
		body["events"] = events
	}
	var w WebhookEndpoint
	if err := s.client.request(ctx, "POST", "/v1/webhooks", body, &w); err != nil {
		return nil, err
	}
	return &w, nil
}

// List returns the account's endpoints (each Secret is masked).
func (s *WebhooksService) List(ctx context.Context) (*WebhookList, error) {
	var out WebhookList
	if err := s.client.request(ctx, "GET", "/v1/webhooks", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Test sends a webhook.test ping to an endpoint and reports the delivery result.
func (s *WebhooksService) Test(ctx context.Context, id string) (*WebhookTestResult, error) {
	var out WebhookTestResult
	if err := s.client.request(ctx, "POST", "/v1/webhooks/"+url.PathEscape(id)+"/test", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Deliveries returns the recent delivery history plus aggregate health for an
// endpoint. limit caps the number of items; when limit <= 0 the server default
// (50) is used and no limit parameter is sent.
func (s *WebhooksService) Deliveries(ctx context.Context, endpointID string, limit int) (*WebhookDeliveriesResult, error) {
	q := url.Values{}
	if limit > 0 {
		q.Set("limit", strconv.Itoa(limit))
	}
	var out WebhookDeliveriesResult
	if err := s.client.request(ctx, "GET", "/v1/webhooks/"+url.PathEscape(endpointID)+"/deliveries"+queryParams(q), nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Replay re-queues a delivery by its event id (e.g. after fixing a receiver).
// The event is delivered again to every endpoint subscribed to it.
func (s *WebhooksService) Replay(ctx context.Context, eventID string) (*ReplayResult, error) {
	var out ReplayResult
	if err := s.client.request(ctx, "POST", "/v1/webhooks/deliveries/"+url.PathEscape(eventID)+"/replay", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Delete removes an endpoint (HTTP 204).
func (s *WebhooksService) Delete(ctx context.Context, id string) error {
	return s.client.request(ctx, "DELETE", "/v1/webhooks/"+url.PathEscape(id), nil, nil)
}

// DeliveryID returns the value of the X-VideoFetch-Delivery header, the stable
// idempotency key for a webhook delivery (it equals the event id). Deliveries
// are at-least-once, so the same delivery id can arrive more than once — record
// it and skip work you already did before processing the event. The lookup is
// case-insensitive; it returns "" when the header is absent.
func DeliveryID(headers http.Header) string {
	return headers.Get("X-VideoFetch-Delivery")
}

// AttemptNumber returns the 1-based delivery attempt from the
// X-VideoFetch-Attempt header (1..3, the first try is 1). It returns 0 when the
// header is absent or not a valid non-negative integer. Deliveries are
// at-least-once and not ordered, so use this only for diagnostics/logging, never
// as an ordering or exactly-once guarantee (order by the seq field instead).
func AttemptNumber(headers http.Header) int {
	n, err := strconv.Atoi(strings.TrimSpace(headers.Get("X-VideoFetch-Attempt")))
	if err != nil || n < 0 {
		return 0
	}
	return n
}

// VerifyWebhookSignature checks the X-VideoFetch-Signature header
// (format: "sha256=<hex digest of raw body>") against the raw request body,
// then parses and returns the event.
func VerifyWebhookSignature(rawBody []byte, sigHeader, secret string) (*WebhookPayload, error) {
	if sigHeader == "" {
		return nil, errors.New("videofetch: no signature header was present")
	}
	if !strings.HasPrefix(sigHeader, "sha256=") {
		return nil, errors.New("videofetch: signature header is malformed (expected sha256=<hex>)")
	}

	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write(rawBody)
	expected := "sha256=" + hex.EncodeToString(mac.Sum(nil))
	if !hmac.Equal([]byte(expected), []byte(sigHeader)) {
		return nil, ErrBadSignature
	}

	var ev WebhookPayload
	if err := json.Unmarshal(rawBody, &ev); err != nil {
		return nil, errors.New("videofetch: payload is not valid JSON")
	}
	return &ev, nil
}

File: typescript/src/webhooks.ts
/**
 * Webhook signature verification.
 *
 * The server signs the raw request body with your endpoint secret (HMAC-SHA256)
 * and sends it in the `X-VideoFetch-Signature` header:
 *
 *     X-VideoFetch-Signature: sha256=<hex digest of raw body>
 *
 * Usage (Next.js route handler):
 *
 *     import { constructEvent } from "@videofetch/sdk";
 *     const payload = await request.text();
 *     const event = constructEvent(payload, request.headers.get("x-videoFetch-signature"), secret);
 */

import { VideoFetchError } from "./errors";
import type { VideoFetch } from "./client";
import type {
  WebhookEndpoint, WebhookEvent, WebhookList, WebhookPayload, WebhookTestResult,
  WebhookDeliveriesResult, WebhookReplayResult,
} from "./types";

export class SignatureVerificationError extends VideoFetchError {
  name = "SignatureVerificationError";
}

/**
 * Compute the expected signature header value for a payload.
 * Pure helper — same as the server side.
 */
export async function computeSignature(payload: string | Uint8Array, secret: string): Promise<string> {
  const subtle = await getSubtle();
  const key = await subtle.importKey(
    "raw",
    new TextEncoder().encode(secret),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"],
  );
  const data = typeof payload === "string" ? new TextEncoder().encode(payload) : payload;
  const sig: ArrayBuffer = await subtle.sign("HMAC", key, data as unknown as BufferSource);
  return `sha256=${toHex(new Uint8Array(sig))}`;
}

// DOM SubtleCrypto (browser/Node19+) vs node:crypto webcrypto are structurally
// incompatible in TS; return as `any` and let the runtime decide which exists.
// eslint-disable-next-line @typescript-eslint/no-explicit-any
async function getSubtle(): Promise<any> {
  if (globalThis.crypto?.subtle) return globalThis.crypto.subtle;
  // Node 18 (before global webcrypto) — dynamic import keeps browser bundles safe
  const { webcrypto } = await import("node:crypto");
  return webcrypto.subtle;
}

/** Verify X-VideoFetch-Signature and return the parsed event payload. */
export async function constructEvent(
  payload: string | Uint8Array,
  sigHeader: string | null | undefined,
  secret: string,
): Promise<WebhookPayload> {
  // The signature is computed over the RAW request body. Passing an already
  // parsed object (e.g. `await req.json()`) loses the exact bytes and would
  // otherwise crash with a native TypeError inside TextDecoder — surface a
  // clear, catchable SignatureVerificationError instead.
  if (typeof payload !== "string" && !(payload instanceof Uint8Array)) {
    throw new SignatureVerificationError(
      "Payload must be the raw request body (string or Uint8Array), not a parsed object. " +
      "Read it as text first, e.g. `await request.text()`.",
    );
  }
  if (!sigHeader) {
    throw new SignatureVerificationError("No signature header was present.");
  }
  if (!sigHeader.startsWith("sha256=")) {
    throw new SignatureVerificationError("Signature header is malformed (expected sha256=<hex>).");
  }
  const expected = await computeSignature(payload, secret);
  if (!timingSafeEqual(expected, sigHeader)) {
    throw new SignatureVerificationError("Signature does not match the payload and secret.");
  }
  try {
    return JSON.parse(typeof payload === "string" ? payload : new TextDecoder().decode(payload)) as WebhookPayload;
  } catch {
    throw new SignatureVerificationError("Payload is not valid JSON.");
  }
}

function toHex(bytes: Uint8Array): string {
  let out = "";
  for (const b of bytes) out += b.toString(16).padStart(2, "0");
  return out;
}

/** Constant-time compare for hex strings (length-equalized). */
function timingSafeEqual(a: string, b: string): boolean {
  const maxLen = Math.max(a.length, b.length);
  let diff = a.length ^ b.length;
  for (let i = 0; i < maxLen; i++) {
    diff |= (a.charCodeAt(i) || 0) ^ (b.charCodeAt(i) || 0);
  }
  return diff === 0;
}

/** Case-insensitive header lookup; returns the first value for multi-value headers. */
function headerValue(
  headers: Record<string, string | string[] | undefined>,
  name: string,
): string | null {
  const target = name.toLowerCase();
  for (const key of Object.keys(headers)) {
    if (key.toLowerCase() !== target) continue;
    const raw = headers[key];
    const value = Array.isArray(raw) ? raw[0] : raw;
    return value == null ? null : value;
  }
  return null;
}

/**
 * Read the delivery id from a webhook request (`X-VideoFetch-Delivery`).
 *
 * The value is the event id (`evt_…`) and is stable across every delivery of
 * the same event. Delivery is **at-least-once**: the same event may be sent
 * more than once (retries and manual replays), and the order of events is not
 * guaranteed (use `WebhookDelivery.seq` to order them). De-duplicate on this
 * value — e.g. remember handled ids and make a repeat a no-op — so consumers
 * stay idempotent.
 *
 * The header name is matched case-insensitively. Returns null when absent.
 */
export function getDeliveryId(headers: Record<string, string | string[] | undefined>): string | null {
  return headerValue(headers, "X-VideoFetch-Delivery");
}

/**
 * Read the delivery attempt number from a webhook request (`X-VideoFetch-Attempt`).
 *
 * The first delivery is attempt 1; retries increment it (up to `max_attempts`,
 * currently 3). Combined with {@link getDeliveryId}, use this to observe retry
 * behaviour — the same delivery id arrives more than once under at-least-once
 * semantics. The header name is matched case-insensitively. Returns null when
 * the header is absent or not an integer.
 */
export function getAttemptNumber(headers: Record<string, string | string[] | undefined>): number | null {
  const raw = headerValue(headers, "X-VideoFetch-Attempt");
  if (raw == null || raw.trim() === "") return null;
  const n = Number(raw);
  return Number.isInteger(n) ? n : null;
}

/**
 * Account-level webhook endpoint management.
 *
 * These endpoints are authenticated by API key (no JWT needed), so a backend
 * integration can self-provision callbacks:
 *
 *   const wh = await client.webhooks.create("https://example.com/hook"); // secret shown once
 *   await client.webhooks.list();                                        // secret is masked
 *   await client.webhooks.test(wh.id);                                   // ping + verify
 *   await client.webhooks.delete(wh.id);                                 // 204
 */
export class WebhooksResource {
  constructor(private client: VideoFetch) {}

  /**
   * POST /v1/webhooks — register an endpoint. Returns the FULL plaintext
   * `secret` (this is the only time it is ever returned; store it now).
   * `events` defaults to every event on the server side.
   */
  async create(url: string, events?: WebhookEvent[]): Promise<WebhookEndpoint> {
    const body: Record<string, unknown> = { url };
    if (events && events.length > 0) body.events = events;
    return this.client.request<WebhookEndpoint>("POST", "/v1/webhooks", { json: body });
  }

  /** GET /v1/webhooks — list endpoints; each `secret` is masked. */
  async list(): Promise<WebhookList> {
    return this.client.request<WebhookList>("GET", "/v1/webhooks");
  }

  /** POST /v1/webhooks/{id}/test — send a `webhook.test` ping to the endpoint. */
  async test(id: string): Promise<WebhookTestResult> {
    return this.client.request<WebhookTestResult>(
      "POST", `/v1/webhooks/${encodeURIComponent(id)}/test`,
    );
  }

  /** DELETE /v1/webhooks/{id} — 204 No Content. */
  async delete(id: string): Promise<void> {
    await this.client.request("DELETE", `/v1/webhooks/${encodeURIComponent(id)}`);
  }

  /**
   * GET /v1/webhooks/{endpoint_id}/deliveries — recent delivery history plus
   * endpoint health: `attempts`, `last_status_code`, `next_attempt_at`, the
   * backoff `retry_schedule_seconds` and `consecutive_failures`.
   *
   * `limit` is the page size (the server default applies when omitted). The
   * endpoint must belong to the caller's account.
   */
  async deliveries(endpointId: string, limit?: number): Promise<WebhookDeliveriesResult> {
    return this.client.request<WebhookDeliveriesResult>(
      "GET",
      `/v1/webhooks/${encodeURIComponent(endpointId)}/deliveries`,
      { params: { limit } },
    );
  }

  /**
   * POST /v1/webhooks/deliveries/{event_id}/replay — re-queue one recorded
   * delivery.
   *
   * The replay reuses the original `event_id` and is at-least-once, so a
   * consumer that de-duplicates on `X-VideoFetch-Delivery` treats an already
   * handled event as a no-op.
   */
  async replay(eventId: string): Promise<WebhookReplayResult> {
    return this.client.request<WebhookReplayResult>(
      "POST",
      `/v1/webhooks/deliveries/${encodeURIComponent(eventId)}/replay`,
    );
  }
}
